### PR TITLE
WIP: STC baseline correction

### PR DESCRIPTION
**Pull request by olafhauk**
Thanks for contributing a pull request! Please make sure you have read the
[contribution guidelines](https://mne.tools/dev/install/contributing.html)
before submitting.

Please be aware that we are a loose team of volunteers so patience is
necessary. Assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Again, thanks for contributing!

#### Reference issue
#8232

#### What does this implement/fix?
Baseline-correction for source estimates is recommended/necessary for some processing steps. For example, for unsigned source estimates, differences in noise levels may result in differences in intensities (e.g. for conditions with very different numbers of trials). This can be compenstated for by baseline-correction.
I will add a method appy_baseline() to source_estimate.py, analogous to evoked.py. I will add documentation to highlight the caveats of baseline correction for source estimates.

#### Additional information
Any additional information you think is important.

### CODE REVIEW

File: mne/tests/test_source_estimate.py [hunk 1]
```diff
@@ -77,6 +77,50 @@
...
+    stcs.append(read_source_estimate(fname_vol, 'sample'))
+
+    # test on different "baseline" intervals
+    baselines = [(0., 0.1), (None, None)]
+
+    for stc in stcs:
+
+        times = stc.times
+
+        for baseline in baselines:
```
Comment (jona-sassenhagen) [inline mne/tests/test_source_estimate.py:95]: I would do

```suggestion
        for (start, stop) in baselines:
```

File: mne/tests/test_source_estimate.py [hunk 2]
```diff
@@ -77,6 +77,50 @@
 rng = np.random.RandomState(0)
 
 
+@testing.requires_testing_data
+def test_stc_baseline_correction():
+    """Test baseline correction for source estimate objects."""
+    # test on different source estimates
+    stcs = []
+    stcs.append(read_source_estimate(fname_stc))
+    stcs.append(read_source_estimate(fname_vol, 'sample'))
```
Comment (jona-sassenhagen) [inline mne/tests/test_source_estimate.py:86]:
```suggestion
   srcs = [read_source_estimate(fname_stc), read_source_estimate(fname_vol, 'sample')]
```

File: mne/tests/test_source_estimate.py [hunk 3]
```diff
@@ -77,6 +77,50 @@
...
+        for baseline in baselines:
+
+            if baseline[0] is None:
+                t0 = stc.times[0]
+            else:
+                t0 = baseline[0]
+            if baseline[1] is None:
+                t1 = stc.times[-1]
+            else:
+                t1 = baseline[1]
```
Comment (jona-sassenhagen) [inline mne/tests/test_source_estimate.py:104]: I would do `t1 = baseline or stc.times[-1]`

File: mne/tests/test_source_estimate.py [hunk 4]
```diff
@@ -77,6 +77,50 @@
...
+
+            # indec for baseline interval (include boundary latencies)
```
Comment (jona-sassenhagen) [inline mne/tests/test_source_estimate.py:106]:
```suggestion
            # index for baseline interval (include boundary latencies)
```

**Comment by olafhauk** [comment]
@jona-sassenhagen Thanks a lot - I hope I managed to implement all your suggestions.

File: mne/utils/docs.py [hunk 5]
```diff
@@ -1789,6 +1789,20 @@
...
+
+    1. Calculate the mean signal of the baseline period.
+    2. Subtract this mean from the **entire** source estimate data.
+
+    .. note:: Baseline correction is appropriate when signal and noise are
+              approximately additive, and the noise level can be estimated from
+              the baseline interval. This can be the case for non-normalized
+              source activities (e.g. signed and unsigned MNE), but it is not
+              the case for normalized estimates (e.g. signal-to-noise ratios,
+              dSPM, sLORETA).
```
Comment (hoechenberger) [inline mne/utils/docs.py:1804]: I think you may need to insert an empty line here, then the doc build will probably work.

Comment (olafhauk) [reply mne/utils/docs.py:1804]: The other notes I looked at didn't have a blank line at the end. Is there a logic to it?

**Comment by olafhauk** [comment]
I don't think the remaining errors have anything to do with me - please let me know.

**Comment by GuillaumeFavelier** [comment]
I restarted Azure and it's green now.

**Comment by hoechenberger** [comment]
@olafhauk You need to mark this PR as ready for review so it can be merged

**Comment by hoechenberger** [comment]
Docs look good: https://23142-1301584-gh.circle-artifacts.com/0/dev/generated/mne.SourceEstimate.html?highlight=sourceestimate#mne.SourceEstimate.apply_baseline

**Review by hoechenberger — APPROVED** [review]

**Comment by hoechenberger** [comment]
@jona-sassenhagen Merge if you're happy

**Comment by agramfort** [comment]
was missing a what's new entry...

@olafhauk can you make a PR to get credit for this work?

**Pull request merged**